src/gridfs/upload.ts

```diff
@@ -147,8 +153,23 @@ export class GridFSBucketWriteStream extends Writable {
    * The stream is considered constructed when the indexes are done being created
    */
   override _construct(callback: (error?: Error | null) => void): void {
-    if (this.bucket.s.checkedIndexes) {
-      return process.nextTick(callback);
+    if (!this.bucket.s.calledOpenUploadStream) {
+      this.bucket.s.calledOpenUploadStream = true;
+
+      checkIndexes(this).then(
+        () => {
+          this.bucket.s.checkedIndexes = true;
+          this.bucket.emit('index');
+        },
+        error => {
+          if (error instanceof MongoOperationTimeoutError) {
+            handleError(this, error, callback);
+          } else {
+            squashError(error);
+            process.nextTick(callback);
+          }
+        }
+      );
     }
     this.bucket.once('index', callback);
```

Comment:
Can we call the callback in the promise resolution now? I think this indirection made sense when we needed to link the promise started in the CTOR to this, but now it is all here, we can avoid setting a listener.